TShockLauncher/Program.cs

```diff
@@ -0,0 +1,66 @@
+﻿/*
+TShock, a server mod for Terraria
+Copyright (C) 2021 Pryaxis & TShock Contributors
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * The purpose of this project is to be the launcher of the TSAPI server.
+ * We use this project:
+ *	- to copy/move around TShockAPI.dll (the TShock plugin to TSAPI)
+ *	- to publish TShock releases.
+ *	- move dependencies to a ./bin folder
+ * 
+ * The assembly name of this launcher (TShock.exe) was decided on by a community poll.
+ */
+
+using System.Reflection;
+
+Dictionary<string, Assembly> _cache = new Dictionary<string, Assembly>();
+
+System.Runtime.Loader.AssemblyLoadContext.Default.Resolving += Default_Resolving;
+
+Start();
+
+/// <summary>
+/// Resolves a module from the ./bin folder, either with a .dll by preference or .exe
+/// </summary>
+Assembly? Default_Resolving(System.Runtime.Loader.AssemblyLoadContext arg1, AssemblyName arg2)
+{
+	if (arg2?.Name is null) return null;
+	if (_cache.TryGetValue(arg2.Name, out Assembly? asm) && asm is not null) return asm;
+
+	var loc = Path.Combine(Environment.CurrentDirectory, "bin", arg2.Name + ".dll");
```

Comment:
`Environment.CurrentDirectory` is the directory from which the application was started, which could be a batch file or startup script that is not home to the intended directory.

```suggestion
	var loc = Path.Combine(Path.GetDirectoryName(%(execasm)%.Location), "bin", arg2.Name + ".dll");
```

is not a perfect solution perhaps, but it will make sure that it grabs assemblies from the intended directory even when a startup file is housed elsewhere

Reply:
im open to the suggetion, but this would likely need testing, mainly for the bundled release variant as sometimes this will have have an invalid .Location property.

Reply:
+1 to rosen's comment